README.org

* tsr-training-data - Tool to generate randomized VOC PASCAL training data 
  :PROPERTIES:
  :TOC:      :include descendants :depth 2
  :END:

Generate pseudorandomized training data in VOC format for traffic sign
detection system.  The images are intented to train YOLOV3 algorithm
implementation in [[https://github.com/zzh8829/yolov3-tf2][zzh8829/yolov3-tf2]].

:CONTENTS:
- [[#architecture][Architecture]]
- [[#getting-help][Getting help]]
- [[#using-create-datapy][Using create-data.py]]
  - [[#generate-example-test-images][Generate example test images]]
  - [[#input-class-images-option-classimages][Input: Class images option classImages]]
  - [[#input-background-images-option-backgrounds][Input: Background images option backgrounds]]
  - [[#output-list-of-test-images-option-imagelist][Output: List of test images option imagelist]]
  - [[#output-list-of-class-names-option-classes][Output: list of class names option classes]]
  - [[#output-generated-test-images-options-images][Output: generated test images options images]]
  - [[#output-generated-xml--annotations-options-annotations][Output: generated XML- annotations options annotations]]
  - [[#convert-images-to-tensorflow-format][Convert images to tensorflow format]]
  - [[#visualy-validate-tensorflow-conversion][Visualy validate tensorflow conversion]]
- [[#filters][Filters]]
  - [[#blur][Blur]]
  - [[#brightness][Brightness]]
  - [[#gamma][Gamma]]
  - [[#resize][Resize]]
  - [[#rotate][Rotate]]
  - [[#perspective][Perspective]]
  - [[#combined-filter][Combined filter]]
:END:

** Architecture

The picture below gives on overview of =tsr-training-data=:

#+name: process
#+name: architecture
#+BEGIN_SRC plantuml :eval no-export :exports results :file pics/architecture.jpg
  node  "zzh8829/yolov3-tf2" as YoloV3Tf2 <<github>> { 

  }

  node  "tsr-training-data" as TsrTrainingData {


        folder backgrounds <<binary>>



         folder "Class images" as classimages <<binary>>
         artifact filters
         component "create-data.py" as createTrainingData

         filters -->  createTrainingData : randomize


      folder out {

         folder tstVOC {
                folder images <<binary>>
                folder annotations <<VOC XML>>
                file classes <<text>>
                file imagelist <<text>>
         }
      }
      component  "marcus2002/yolov3-tf2-training" as Marcus2002 <<Docker>>


    folder tfData {

      file val.tf <<tensorflow data>>
      file train.tf  <<tensorflow data>>
    }

    actor "Visual validation" as jpgval


  }


      classimages --> createTrainingData
      backgrounds --> createTrainingData


      createTrainingData --> images 
      createTrainingData --> annotations
      createTrainingData --> classes
      createTrainingData --> imagelist



  YoloV3Tf2 .> Marcus2002 : Dockerized

  images --> Marcus2002
  annotations --> Marcus2002
      classes --> Marcus2002
      imagelist --> Marcus2002 : split into two\nfor val and train images


  Marcus2002 --> val.tf : create
  Marcus2002 --> train.tf : create

  val.tf --> jpgval : extract random picture
  train.tf --> jpgval : extract random picture

  node  "yolov3 tf2 training" as Marcus2002.2
  tfData .> Marcus2002.2 : for training ylov3


  #+END_SRC

  #+RESULTS: architecture
  [[file:pics/architecture.jpg]]

=tsr-training-data= merges randomly filtered /class images/ and
/background images/ to create VOC Pascal dataset, which can be
converted to tensforflow format and visually inspected using
dockerized tools from [[https://github.com/zzh8829/yolov3-tf2][zzh8829/yolov3-tf2]]. After validation tensorflow
format test images are passed further for yolo3 training.


** Getting help

Running  =create-data.py= with =--help= option 

#+name: usage
#+BEGIN_SRC sh :eval no-export :results output :exports both
python  create-data.py  --help
#+END_SRC

shows options

#+RESULTS: usage
#+begin_example

       USAGE: create-data.py [flags]
flags:

create-data.py:
  --annotations: path to annotations folder
    (default: 'out/tsrVOC/Annotations')
  --backgrounds: path to backgroud images
    (default: 'backgrounds')
  --blur: Blur filters used on classImages;
    repeat this option to specify a list of values
    (default: '[3, 5]')
    (an integer)
  --blur2: Blur filters used on test images;
    repeat this option to specify a list of values
    (default: '[1, 2, 3]')
    (an integer)
  --brightness: Brigness filters used on classImages;
    repeat this option to specify a list of values
    (default: '[-70, -20]')
    (an integer)
  --brightness2: Brigness filters used on test images;
    repeat this option to specify a list of values
    (default: '[-30, -20]')
    (an integer)
  --classImageWrangles: Filter wrangles to run on class images;
    repeat this option to specify a list of values
    (default: "['blur', 'rotate', 'brightness', 'perspective']")
  --classImages: path to class images
    (default: 'Images/signs')
  --classes: path file where classnames are written
    (default: 'out/tsrVOC/classes.txt')
  --classesTemplate: template to interpolate classes file lines
    (default: '{value}-{startOrEnd}')
  --debug: -3=fatal, -1=warning, 0=info, 1=debug
    (default: '-1')
    (an integer)
  --imagelist: path file where list of images are written
    (default: 'out/tsrVOC/imagelist.txt')
  --imagelistTemplate: template to interpolate imagelist file lines
    (default: '{imageFileBasename} -1\n')
  --images: output folder where test images are written
    (default: 'out/tsrVOC/JPEGImages')
  --[no]interactive: show training data
    (default: 'false')
  --maxImages: Number of images to generate
    (default: '7')
    (an integer)
  --perspective: Possible perspective values;
    repeat this option to specify a list of values
    (default: "['45,-10', '10,10', '-30,10']")
  --rotates: Rotate filters used on classImages;
    repeat this option to specify a list of values
    (default: '[10, 4, -5, -10]')
    (an integer)
  --seed: Random seed (default None)
    (an integer)
  --testImageWrangles: Filter wrangles to run on test images (=final result);
    repeat this option to specify a list of values
    (default: "['brightness', 'blur']")
  --testSet: Test set name (=folder in output -directory)
    (default: 'TSR1')

Try --helpfull to get a list of all flags.
#+end_example



** Using =create-data.py=

#+BEGIN_SRC sh :eval no-export :results output :exports none
rm -rf out/tsrVOC
#+END_SRC

#+RESULTS:

*** Generate example test images 

Running =create-data.py= without any options generates example test
images.  The option =--seed 42= is used only to reproduce execution
results for the example.  For a more realistic example refer to
[[file:yolov3-data.org]].


#+name: run-default
#+BEGIN_SRC sh :eval no-export :results output :exports both
python  create-data.py --seed 42
#+END_SRC

The output explains the generation run

#+RESULTS: run-default
: Created 7 test images in 'out/tsrVOC/JPEGImages' with annotations in 'out/tsrVOC/Annotations' folders
:  class names in 'out/tsrVOC/classes.txt' and list of images 'out/tsrVOC/imagelist.txt'
:  input from from class images 'Images/signs' and backgrounds'backgrounds' folders


*** Input: Class images option =classImages= 

Class image directory is pointed by =classImages= option. 

Option =classesTemplate= defines template, which controls, how class
name is extracted from image names. This template referes to names in
naming pattern =value-type-startOrEnd.ext=.


Default class images located in directory =Images/signs= take
following values:
- =value=  : ={20,30,40,50,60,70,80,100,120,urban}=
- =type= :  ={sign, area,led}=
- =startOrEnd= :  ={start,end}=
- =ext= is valid image file extension e.g. =png=, =jpg=


#+BEGIN_SRC sh :eval no-export :results output :exports results
ls -ltr Images/signs
#+END_SRC

#+RESULTS:
#+begin_example
total 3612
-rw-rw-r-- 1 jj jj 176685 touko  3 19:58 20-sign-start.png
-rw-rw-r-- 1 jj jj 148167 touko  3 19:59 70-sign-start.png
-rw-rw-r-- 1 jj jj 159586 touko  3 19:59 80-sign-start.png
-rw-rw-r-- 1 jj jj 159237 touko  3 20:00 100-sign-start.png
-rw-rw-r-- 1 jj jj 159774 touko  3 20:00 120-sign-start.png
-rw-rw-r-- 1 jj jj 139417 touko  3 20:01 20-sign-end.png
-rw-rw-r-- 1 jj jj 138868 touko  3 20:02 40-sign-end.png
-rw-rw-r-- 1 jj jj 141187 touko  3 20:03 60-sign-end.png
-rw-rw-r-- 1 jj jj  91977 touko  3 20:03 30-area-start.png
-rw-rw-r-- 1 jj jj  84133 touko  3 20:04 40-area-start.png
-rw-rw-r-- 1 jj jj  90324 touko  3 20:04 50-area-start.png
-rw-rw-r-- 1 jj jj  92117 touko  3 20:04 60-area-start.png
-rw-rw-r-- 1 jj jj 144065 touko  3 20:05 30-area-end.png
-rw-rw-r-- 1 jj jj 142809 touko  3 20:05 40-area-end.png
-rw-rw-r-- 1 jj jj 144341 touko  3 20:05 50-area-end.png
-rw-rw-r-- 1 jj jj 145483 touko  3 20:06 60-area-end.png
-rw-rw-r-- 1 jj jj  38379 touko  3 20:14 80-led-start.png
-rw-rw-r-- 1 jj jj  60020 touko  3 20:15 120-led-start.png
-rw-rw-r-- 1 jj jj   7417 touko  3 20:16 100-led-start.png
-rw-rw-r-- 1 jj jj 167258 touko  4 10:34 50-sign-start.png
-rw-rw-r-- 1 jj jj 139883 touko  4 10:36 50-sign-end.png
-rw-rw-r-- 1 jj jj 140155 touko  4 10:49 30-sign-end.png
-rw-rw-r-- 1 jj jj 156815 touko  4 10:50 30-sign-start.png
-rw-rw-r-- 1 jj jj 148650 touko  4 10:51 40-sign-start.png
-rw-rw-r-- 1 jj jj 132627 touko  4 10:52 70-sign-end.png
-rw-rw-r-- 1 jj jj 210360 touko  4 10:56 60-sign-start.png
-rw-rw-r-- 1 jj jj  47325 touko  4 10:59 urban-area-start.png
-rw-rw-r-- 1 jj jj  90642 touko  4 11:00 urban-area-end.png
-rw-rw-r-- 1 jj jj  42307 touko  4 13:10 30-led-start.png
-rw-rw-r-- 1 jj jj  43540 touko  4 13:11 60-led-start.png
-rw-rw-r-- 1 jj jj  44196 touko  4 13:13 50-led-start.png
#+end_example

For example, one of the class images =100-sign-start.png= 

[[file:Images/signs/100-sign-start.png]]


*** Input: Background images option =backgrounds=

=create-data.py= merges class images with background images under
directory pointed by =backgrounds= -option. Example backaground images
packaged with the tool are:

#+BEGIN_SRC sh :eval no-export :results output :exports results
find backgrounds \( -name '*.JPG' -o  -name '*.jpg' \)
#+END_SRC

#+RESULTS:
: backgrounds/misc/frilly_0032.jpg
: backgrounds/misc/frilly_0003.jpg
: backgrounds/roads/CIMG0726.JPG.jpg
: backgrounds/roads/CIMG3009.JPG.jpg
: backgrounds/forest/SAM_0807.JPG
: backgrounds/forest/SAM_0808.JPG
: backgrounds/forest/SAM_0806.JPG

One of these examples, =backgrounds/roads/CIMG0726.JPG.jpg=, is shown
below:

[[file:backgrounds/roads/CIMG0726.JPG.jpg]]


*** Output: List of test images option =imagelist=

Names of generated test image files are written into file pointed by
=imagelist= option. Format of the line is given by =imagelistTemplate=
-option.

For the example the tool outputs
#+BEGIN_SRC sh :eval no-export :results output :exports results
cat out/tsrVOC/imagelist.txt
#+END_SRC

#+RESULTS:
: TSR-image00000 -1
: TSR-image00001 -1
: TSR-image00002 -1
: TSR-image00003 -1
: TSR-image00004 -1
: TSR-image00005 -1
: TSR-image00006 -1


*** Output: list of class names option =classes=

Class names of the images are written into a file pointed by =classes=
options. 

For the example run class names are:

#+BEGIN_SRC sh :eval no-export :results output :exports results
cat out/tsrVOC/classes.txt
#+END_SRC

#+RESULTS:
: 40-end
: 50-start
: 60-end
: 30-start
: 60-start
: urban-start
: 30-end




*** Output: generated test images options =images=

Test images are genered into the directory pointed by =images= option. 

For the example, the tool generates the following files

 #+BEGIN_SRC sh :eval no-export :results output :exports results 
 ls -tr out/tsrVOC/JPEGImages/
 #+END_SRC

 #+RESULTS:
 : TSR-image00000.jpg
 : TSR-image00001.jpg
 : TSR-image00002.jpg
 : TSR-image00003.jpg
 : TSR-image00004.jpg
 : TSR-image00005.jpg
 : TSR-image00006.jpg

One of the test images =TSR-image00002.jpg=:

 #+BEGIN_SRC sh :eval no-export :results output raw :exports results
 find out/tsrVOC/JPEGImages -name '*002.jpg' -exec echo [[file:{}]] \;
 #+END_SRC

 #+RESULTS:
 [[file:out/tsrVOC/JPEGImages/TSR-image00002.jpg]]



*** Output: generated XML- annotations options =annotations=

Test image annotatios are generated into directory pointed by
=annotations= -option.  Below is the content for example run:

 #+BEGIN_SRC sh :eval no-export :results output :exports results
 ls -ltr out/tsrVOC/Annotations/
 #+END_SRC

 #+RESULTS:
 : total 28
 : -rw-rw-r-- 1 jj jj 606 touko 13 10:48 TSR-image00000.xml
 : -rw-rw-r-- 1 jj jj 603 touko 13 10:48 TSR-image00001.xml
 : -rw-rw-r-- 1 jj jj 601 touko 13 10:48 TSR-image00002.xml
 : -rw-rw-r-- 1 jj jj 605 touko 13 10:48 TSR-image00003.xml
 : -rw-rw-r-- 1 jj jj 607 touko 13 10:48 TSR-image00004.xml
 : -rw-rw-r-- 1 jj jj 606 touko 13 10:48 TSR-image00005.xml
 : -rw-rw-r-- 1 jj jj 605 touko 13 10:48 TSR-image00006.xml


Example annotation file =TSR-image00002.xml=

 #+BEGIN_SRC sh :eval no-export :results output :exports results
 cat out/tsrVOC/Annotations/TSR-image00002.xml
 #+END_SRC

 #+RESULTS:
 #+begin_example
 <annotation>
         <folder>TSR1</folder>
         <filename>TSR-image00002.jpg</filename>
         <source>
                 <database>TSR training data</database>
                 <annotation>classInfo: {'value': '20', 'type': 'sign', 'startOrEnd': 'start'}</annotation>
                 <image>flickr</image>
         </source>
         <size>
                 <width>1200</width>
                 <height>900</height>
                 <depth>3</depth>
         </size>
         <segmented>0</segmented>
         <object>
                 <name>20-start</name>
                 <pose>Unspecified</pose>
                 <truncated>0</truncated>
                 <difficult>0</difficult>
                 <bndbox>
                         <xmin>268</xmin>
                         <ymin>161</ymin>
                         <xmax>499</xmax>
                         <ymax>390</ymax>
                 </bndbox>
         </object>
 </annotation>
 #+end_example




*** Convert images to tensorflow format


 #+BEGIN_SRC sh :eval no-export :results output :exports none
 mkdir out/tfData
 #+END_SRC


**** Split image set to training and validation


 #+RESULTS:


 Split =out/tsrVOC/imagelist.txt= into two files
 =out/tsrVOC/ImageSets/Main/aeroplane_train.txt= and
 =out/tsrVOC/ImageSets/Main/aeroplane_val.txt= used by [[https://github.com/zzh8829/yolov3-tf2/blob/master/tools/voc2012.py][yolov3-tf2 VOC
 conversion tools]]

 #+BEGIN_SRC sh :eval no-export :results output
 mkdir -p out/tsrVOC/ImageSets/Main
 tail -3 out/tsrVOC/imagelist.txt > out/tsrVOC/ImageSets/Main/aeroplane_train.txt
 head  -4 out/tsrVOC/imagelist.txt > out/tsrVOC/ImageSets/Main/aeroplane_val.txt
 #+END_SRC

 #+RESULTS:

 #+BEGIN_SRC sh :eval no-export :results output :exports none
 ls -ltr out/tsrVOC/ImageSets/Main
 #+END_SRC

 #+RESULTS:
 : total 8
 : -rw-rw-r-- 1 jj jj 72 touko 13 14:00 aeroplane_val.txt
 : -rw-rw-r-- 1 jj jj 54 touko 13 14:00 aeroplane_train.txt



**** Convert training dataset to tensorflow format

 #+BEGIN_SRC sh :eval no-export :results output :var TAG=tag-number
   docker run \
        --user $(id -u):$(id -g) \
        --workdir /yolov3-tf2 \
        --volume $(pwd)/out/tsrVOC/:/yolov3-tf2/tsrVOC \
        --volume $(pwd)/out/tfData/:/yolov3-tf2/tfData \
        marcus2002/yolov3-tf2-training:$TAG \
          python tools/voc2012.py \
            --classes tsrVOC/classes.txt \
            --data_dir tsrVOC \
            --output_file tfData/tsr_train.tfrecord \
            --split train
 #+END_SRC



 #+RESULTS:

 The result is

 #+BEGIN_SRC sh :eval no-export :results output :exports results
 ls -tr out/tfData/tsr_train.tfrecord
 #+END_SRC

 #+RESULTS:
 : out/tfData/tsr_train.tfrecord


**** Convert validation dataset to tensorflow format

 #+BEGIN_SRC sh :eval no-export :results output :var TAG=tag-number
   docker run \
        --user $(id -u):$(id -g) \
        --workdir /yolov3-tf2 \
        --volume $(pwd)/out/tsrVOC/:/yolov3-tf2/tsrVOC \
        --volume $(pwd)/out/tfData/:/yolov3-tf2/tfData \
        marcus2002/yolov3-tf2-training:$TAG \
          python tools/voc2012.py \
            --classes tsrVOC/classes.txt \
            --data_dir tsrVOC \
            --output_file tfData/tsr_val.tfrecord \
            --split val
 #+END_SRC

 #+RESULTS:

 The result is

 #+BEGIN_SRC sh :eval no-export :results output :exports results
 ls -tr out/tfData/tsr_val.tfrecord
 #+END_SRC

 #+RESULTS:
 : out/tfData/tsr_val.tfrecord


*** Visualy validate tensorflow conversion

 To visualize training tensorflow data in
 =out/tfData/tsr_train.tfrecord= run the command

 #+BEGIN_SRC sh :eval no-export :results output :var TAG=tag-number
   docker run \
        --user $(id -u):$(id -g) \
        --workdir /yolov3-tf2 \
        --volume $(pwd)/out/tsrVOC/:/yolov3-tf2/tsrVOC \
        --volume $(pwd)/out/tfData/:/yolov3-tf2/tfData \
        marcus2002/yolov3-tf2-training:$TAG \
          python tools/visualize_dataset.py \
            --classes tsrVOC/classes.txt \
            --dataset  tfData/tsr_train.tfrecord \
            --output tfData/visu-trainset.jpg


 #+END_SRC

 #+RESULTS:

 The result shows

 [[file:out/tfData/visu-trainset.jpg]]


 To visualize dataset in =out/tfData/tsr_val.tfrecord= run

 #+BEGIN_SRC sh :eval no-export :results output :var TAG=tag-number
   docker run \
        --user $(id -u):$(id -g) \
        --workdir /yolov3-tf2 \
        --volume $(pwd)/out/tsrVOC/:/yolov3-tf2/tsrVOC \
        --volume $(pwd)/out/tfData/:/yolov3-tf2/tfData \
        marcus2002/yolov3-tf2-training:$TAG \
          python tools/visualize_dataset.py \
            --classes tsrVOC/classes.txt \
            --dataset  tfData/tsr_val.tfrecord \
            --output tfData/visu-valset.jpg
 #+END_SRC

 #+RESULTS:

 The result shows

 [[file:out/tfData/visu-valset.jpg]]


** Filters

This chapter documents filters in =create-data.py= using tables with
columns for
- filter input parameters
- the result of filtering image shown below
- and the associated mask, which filter also produces

The unfiltered 200 pixel wide image 


 #+RESULTS: filtered-image
 [[file:./pics/example.jpg]]


#+BEGIN_SRC python :eval no-export :results output :noweb no :session *Python* :exports none
  for moduleName in [ 'src.imageTools', "src.classImages"]:
      if moduleName  in sys.modules:
          del sys.modules[moduleName]

  import src.util
  import cv2
  import os.path
  import imutils
  import src.imageTools
  import src.classImages

  def imageLink( imagePath ):
      return( "[[file:./" + imagePath + "]]" )

  def filterImage( imagePath, filter, filterMask, imageFile=None, maskFile=None, width =100 ):

      # prepare image && mask for  filtering
      img = cv2.imread( imagePath)
      cropped, mask = src.classImages.maskImage(img)
      if width is not None: 
          img = imutils.resize( img, width=width )
          mask = imutils.resize( mask, width=width )
      img, filteredMask = filter( img, mask=mask )
      ## if filterMask: mask = filter( mask )

      if imageFile is None: imageFile = os.path.basename(imagePath)
      if maskFile is None: maskFile = os.path.basename(imagePath) + "-mask"
      picPath = os.path.join( "pics", imageFile )
      cv2.imwrite( picPath, img )
      maskPath = os.path.join( "pics", maskFile )
      if filteredMask is not None: cv2.imwrite( maskPath, filteredMask )
      return( imageLink(picPath), imageLink(maskPath) )


  def filterDocument( imagePath, filterName, filterTool, filterValues, filterMask=False, strValues=None ):
      def printRow( col1, col2, col3,  sep="|" ):
          if sep is not None: 
              print( sep, col1, sep, col2, sep, col3, sep)
          else:
              print( col1, col2,  )

      printRow( filterName, "Filtered image", "Mask" )
      printRow( "|---+---+---|", "", "", sep = None )

      for index, filterValue in enumerate(filterValues):
          if  strValues is None:
              strValue = str(filterValue)
          else:
              strValue = str( strValues[index])
          strValueInName = strValue.replace( '[', "").replace( ']', "").replace( '(', "").replace( ')', "").replace( ',', "").replace( ' ', "")
          imageFile =  filterName + strValueInName + ".png"
          maskFile = filterName + strValueInName + "-mask.png"
          imageLink, maskLink = filterImage(
              imagePath,
              lambda img, mask=None: filterTool( img, filterValue, mask=mask ), 
              filterMask,
              imageFile = imageFile,
              maskFile = maskFile)

          printRow( strValue, imageLink, maskLink )

      printRow( "|---+---+---|", "", "", sep = None )

#+END_SRC

#+RESULTS:


 #+name: filtered-image
 #+BEGIN_SRC python :eval no-export :results output raw :session *Python* :exports results
   imagePath = "Images/signs/50-sign-start.png"
   filter = lambda img, mask: src.imageTools.resize_image( img, 200, mask=None )
   picPath,_ = filterImage( imagePath, filter, False, imageFile="example.jpg" )
   print(picPath )

 #+END_SRC

#+RESULTS:


*** Blur

Blur filter options are =blur= (for class images) and =blur2= (for merged test image).

 #+BEGIN_SRC python :eval no-export :results output raw :session *Python* :exports results
   imagePath = "Images/signs/50-sign-start.png"
   filterTool = src.imageTools.blur_image
   filterValues = [ 1,3,5,10 ]

   filterDocument( imagePath, "Blur", filterTool, filterValues )
 #+END_SRC

 #+RESULTS:
 | Blur | Filtered image         | Mask                        |
 |------+------------------------+-----------------------------|
 |    1 | [[file:./pics/Blur1.png]]  | [[file:./pics/Blur1-mask.png]]  |
 |    3 | [[file:./pics/Blur3.png]]  | [[file:./pics/Blur3-mask.png]]  |
 |    5 | [[file:./pics/Blur5.png]]  | [[file:./pics/Blur5-mask.png]]  |
 |   10 | [[file:./pics/Blur10.png]] | [[file:./pics/Blur10-mask.png]] |
 |------+------------------------+-----------------------------|


*** Brightness

Blur filter options are =brightness= (for class images) and
=brightness2= (for merged test image).

 #+BEGIN_SRC python :eval no-export :results output raw :session *Python* :exports results
   imagePath = "Images/signs/50-sign-start.png"
   filterTool = src.imageTools.brightness_image
   filterValues = [ -250, -200, -100, -50, 0, 50, 100, 200, 250 ]

   filterDocument( imagePath, "Brightness", filterTool, filterValues )
 #+END_SRC

 #+RESULTS:
 | Brightness | Filtered image                 | Mask                                |
 |------------+--------------------------------+-------------------------------------|
 |       -250 | [[file:./pics/Brightness-250.png]] | [[file:./pics/Brightness-250-mask.png]] |
 |       -200 | [[file:./pics/Brightness-200.png]] | [[file:./pics/Brightness-200-mask.png]] |
 |       -100 | [[file:./pics/Brightness-100.png]] | [[file:./pics/Brightness-100-mask.png]] |
 |        -50 | [[file:./pics/Brightness-50.png]]  | [[file:./pics/Brightness-50-mask.png]]  |
 |          0 | [[file:./pics/Brightness0.png]]    | [[file:./pics/Brightness0-mask.png]]    |
 |         50 | [[file:./pics/Brightness50.png]]   | [[file:./pics/Brightness50-mask.png]]   |
 |        100 | [[file:./pics/Brightness100.png]]  | [[file:./pics/Brightness100-mask.png]]  |
 |        200 | [[file:./pics/Brightness200.png]]  | [[file:./pics/Brightness200-mask.png]]  |
 |        250 | [[file:./pics/Brightness250.png]]  | [[file:./pics/Brightness250-mask.png]]  |
 |------------+--------------------------------+-------------------------------------|


*** Gamma

Not used.

 #+BEGIN_SRC python :eval no-export :results output raw :session *Python* :exports results

ChangeLog
2020-05-20  jj  <jj@santtu>

	* README.org (Running): -seed 42

2020-05-19  jj  <jj@santtu>

	* README.org: fixed image link for perspective and multi filter
	examples
	(Option): documentation added to filters

2020-05-15  jj  <jj@santtu>

	* README.org: added filters



create-data.py
import os.path
import sys
from absl import app, flags, logging
from absl.flags import FLAGS
import random
import cv2

import src.createTrainingData
import src.backgrounds
import src.classImages

# control
flags.DEFINE_integer('debug', -1, '-3=fatal, -1=warning, 0=info, 1=debug')
flags.DEFINE_integer('seed', None, 'Random seed (default None)')
flags.DEFINE_boolean('interactive', False, 'show training data')
flags.DEFINE_integer("maxImages", 7, "Number of images to generate")

# input
flags.DEFINE_string('classImages', "Images/signs", "path to class images")
flags.DEFINE_string('backgrounds', 'backgrounds', 'path to backgroud images')
flags.DEFINE_string('testSet', "TSR1", 'Test set name (=folder in output -directory)')

# class image filters
flags.DEFINE_multi_integer( "rotates", [10,4  ,-5, -10], "Rotate filters used on classImages")
flags.DEFINE_multi_integer( "blur", [3,5], "Blur filters used on classImages")
flags.DEFINE_multi_integer( "brightness", [-70,-20], "Brigness filters used on classImages")
flags.DEFINE_multi_string( "perspective", [ "45,-10", "10,10", "-30,10" ], "Possible perspective values")
flags.DEFINE_multi_string( "classImageWrangles", [ "blur", "rotate", "brightness", "perspective" ], "Filter wrangles to run on class images" )

# test image filters
flags.DEFINE_multi_integer( "brightness2", [-30,-20], "Brigness filters used on test images")
flags.DEFINE_multi_integer( "blur2", [1,2,3], "Blur filters used on test images")
flags.DEFINE_multi_string( "testImageWrangles", [ "brightness", "blur" ], "Filter wrangles to run on test images (=final result)" )

# output
flags.DEFINE_string('images', "out/tsrVOC/JPEGImages", 'output folder where test images are written')
flags.DEFINE_string( 'annotations', "out/tsrVOC/Annotations" , "path to annotations folder")
flags.DEFINE_string('classes', "out/tsrVOC/classes.txt", 'path file where classnames are written')
flags.DEFINE_string( "classesTemplate",
                     "{value}-{startOrEnd}", "template to interpolate classes file lines")

flags.DEFINE_string('imagelist', "out/tsrVOC/imagelist.txt",
                    'path file where list of images are written')
flags.DEFINE_string( "imagelistTemplate",
                     "{imageFileBasename} -1\n", "template to interpolate imagelist file lines")


## flags.DEFINE_list( "rotates", "-32,-10,-5,3,35,45", "rotate angles to wrangle class images")

def ensure_dir(directory):
    if not os.path.exists(directory):
        os.makedirs(directory)

def ensure_filedir(file_path):
    directory = os.path.dirname(file_path)
    if not os.path.exists(directory):
        os.makedirs(directory)        

def main(_argv):


    def progress(count, total, status=''):
        # https://gist.github.com/vladignatyev/06860ec2040cb497f0f3
        bar_len = 60
        filled_len = int(round(bar_len * count / float(total)))

        percents = round(100.0 * count / float(total), 1)
        bar = '=' * filled_len + '-' * (bar_len - filled_len)

        sys.stderr.write('[%s] %s%s ...%s\r' % (bar, percents, '%', status))
        sys.stderr.flush()  # As suggested by Rom Ruben (see: http://stackoverflow.com/questions/3173320/text-progress-bar-in-the-c

    logging.set_verbosity(FLAGS.debug)
    random.seed( FLAGS.seed )
    logging.info( "Random seed {0}".format(FLAGS.seed))

    # Default values
    if FLAGS.annotations is None:
        FLAGS.annotations = os.path.join( FLAGS.images, "annotations")

    
    logging.info('starting')

    def resetBackgroundGen():
        return( src.backgrounds.yieldBackgroundImagePaths(FLAGS.backgrounds) )

    # Generator yielding backgroud images
    backgroundGen = resetBackgroundGen()

    # Paste random classImages to back
    classImagesList = list( src.classImages.yieldCroppedClassImages(FLAGS.classImages) )

    # Set of filters randomly maninpulating classImage before pasting
    # into background
    classImagesFilters = src.imageTools.Filters(
        { "rotate" : FLAGS.rotates, "blur": FLAGS.blur
          , "brightness": FLAGS.brightness
          , "perspective": FLAGS.perspective
    })
    classImageWrangles = classImagesFilters.imageWrangles( wrangleTypes = FLAGS.classImageWrangles)
    logging.info( "classImageWrangles={0}".format( classImageWrangles))

    # Wranlers to run on test images (=finalize result)

    testImagesFilters = src.imageTools.Filters(
        { 
            "brightness": FLAGS.brightness2
            , "blur": FLAGS.blur2
        })    
    testImageWrangles = testImagesFilters.imageWrangles(wrangleTypes = FLAGS.testImageWrangles)
    logging.info( "testImageWrangles={0}".format( testImageWrangles))    
    
    

    logging.info( "Merge class images '{0}' with backgrounds '{1}' to a test set '{2}' with image  '{3}' and annotations {4} folders ".format(
        FLAGS.classImages, FLAGS.backgrounds, FLAGS.testSet, FLAGS.images, FLAGS.annotations ))

    # truncate imagelist file
    imagesDirectory = os.path.join(FLAGS.images)
    annotationDirectory =  os.path.join(FLAGS.annotations)

    # make sure that directories exist 
    ensure_dir( imagesDirectory )
    ensure_dir( annotationDirectory )
    ensure_filedir( FLAGS.imagelist )
    ensure_filedir( FLAGS.classes )    


    # empty imagelist
    open( FLAGS.imagelist, "w").close()
    
    # Collect unique class names
    classnames = set()
    imageCount = 0
    while True:
        
        for mergedImage in  src.createTrainingData.yieldMergedImages(
                    backgroundGen,
                    classImagesList, classImagesFilters,
                    wrangles=classImageWrangles,
                    debug=False, debugDebug=False ):

            if imageCount >= FLAGS.maxImages:
                break


            # iterate 'backgroundGen', choose random image from
            # 'classImagesList', and run random wrangler from
            # 'classImageWrangles' and merge classImage with background

            logging.info( "classInfo value: {0} type: {1} startOrEnd: {2}".format(
                mergedImage["classInfo"]["value"], mergedImage["classInfo"]["type"], mergedImage["classInfo"]["startOrEnd"], mergedImage["classInfo"]) )

            # Use template to define class name
            mergedImage["classname"] = FLAGS.classesTemplate.format(**mergedImage["classInfo"] )

            # actual merge
            testImage, _ = src.createTrainingData.createTestImage(
                mergedImage, imageCount , testSet=FLAGS.testSet,
                wrangles=testImageWrangles, filters=testImagesFilters )

            if FLAGS.interactive: 
                imgBox = src.imageTools.drawBoundingBox( testImage["testImage"], testImage["boundingBox"] )
                src.util.showImage( imgBox, "Test image " + str(testImage["classInfo"] ) )



            # Collect classnames to a set 

            classnames.add(testImage["classname"])

            # write 1) image, 2) label xml,  3) imagelist file entry
            imageCount = imageCount + 1
            cv2.imwrite( os.path.join( imagesDirectory, testImage["imageFilename"]), testImage["testImage"] )
            with open( os.path.join( annotationDirectory, testImage["labelFilename"]), "w+" ) as labelFile:
                labelFile.write( testImage["vocXML"] )
            with open( FLAGS.imagelist, "a") as imagelistFile:
                ## imagelistFile.write( FLAGS.imagelistTemplate.format( ).format( )testImage["imageFilename"] + -1"\n" )
                testImage["imageFileBasename"] = os.path.splitext( os.path.basename( testImage["imageFilename"] ))[0]
                imagelistFileRow = FLAGS.imagelistTemplate.format(**testImage )
                imagelistFile.write( str(imagelistFileRow)  )

            progress( imageCount, FLAGS.maxImages, "Images")

        # all requestesd images done
        if imageCount >= FLAGS.maxImages:
            break
        else:
            # Restart generator yielding backgroud images
            backgroundGen = resetBackgroundGen()



    # create classes file with unique classnames
    classNameLines = [ cName+"\n" for cName in sorted(classnames) ]
    with open( FLAGS.classes, "w+" ) as classFile:
        classFile.writelines( classNameLines )
            
    print( "Created {0} test images in '{1}' with annotations in '{2}' folders".format(
        imageCount, FLAGS.images,  FLAGS.annotations ))
    print( " class names in '{0}' and list of images '{1}'".format(
        FLAGS.classes, FLAGS.imagelist))
    print( " input from from class images '{0}' and backgrounds'{1}' folders".format(
        FLAGS.classImages, FLAGS.backgrounds ))



if __name__ == '__main__':
    try:
        app.run(main)
    except SystemExit:
        pass
    
